### PR TITLE
Shooter sim

### PR DESCRIPTION
**Pull request by vichik123**

### CODE REVIEW

**Comment by vichik123** [comment]
What the hell? I didn't change the libraries.

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -77,6 +78,7 @@ public void robotInit() {
         SignalLogger.enableAutoLogging(true);
 
         ExampleSubsystemConstants.initConstants();
+        ShooterConstants.initialize(Constants.CURRENT_MODE);
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java]: I recommend against making changes to global files before all subsystems are merged, this will cause merge conflicts. Better to not push these changes and after all subsystems are merged, make a small PR with all the initializations.

File: src/main/java/frc/robot/RobotContainer.java
Comment (katzuv) [inline src/main/java/frc/robot/RobotContainer.java:1]: Here too should not be pushed, although you _could_ make unit tests that test the commands if you're adventurous...

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 2]
```diff
@@ -35,21 +36,38 @@ public static void initialize(ShooterIO io) {
...
-        return run(() -> io.setVelocity(velocity.get())).withName("Set Shooter Velocity");
+        return setVelocity(velocity, velocity);
+    }
+
+    public MutableMeasure<Velocity<Angle>> getTopVelocity() {
+        return inputs.topVelocity;
+    }
+
+    public MutableMeasure<Velocity<Angle>> getBottomVelocity() {
+        return inputs.bottomVelocity;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Why are these needed?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 3]
```diff
@@ -35,21 +36,38 @@ public static void initialize(ShooterIO io) {
...
     }
 
-    public MutableMeasure<Velocity<Angle>> getVelocity() {
-        return inputs.velocity;
+    public Command stop() {
+        return setVelocity(() -> Units.RotationsPerSecond.zero().mutableCopy());
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: `Units.RotationsPerSecond.zero().mutableCopy()` should be a constant.

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java [hunk 4]
```diff
@@ -1,6 +1,62 @@
...
+    public static final double GEAR_RATIO = 1.0;
+    public static final double SETPOINT_TOLERANCE = 0.05; // Percentages
+    public static final double MOMENT_OF_INERTIA = 0.08;
+
+    public static final LoggedTunableNumber TOP_P = new LoggedTunableNumber("Shooter/Top P");
+    public static final LoggedTunableNumber TOP_I = new LoggedTunableNumber("Shooter/Top I");
+    public static final LoggedTunableNumber TOP_D = new LoggedTunableNumber("Shooter/Top D");
+    public static final LoggedTunableNumber TOP_Ks = new LoggedTunableNumber("Shooter/Top Ks");
+    public static final LoggedTunableNumber TOP_Kv = new LoggedTunableNumber("Shooter/Top Kv");
+    public static final LoggedTunableNumber TOP_Ka = new LoggedTunableNumber("Shooter/Top Ka");
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: `kS` instead of `Ks`. FYI, with [`VelocityTorqueCurrentFOC`](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/controls/VelocityTorqueCurrentFOC.html) you _generally_ don't need a kV (but I wouldn't delete it):

> Since the torque request is directly proportional to acceleration, $K_v$ is generally unnecessary. A torque output of 0 corresponds to a constant velocity, assuming no external forces.

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java [hunk 5]
```diff
@@ -1,6 +1,62 @@
...
 public class ShooterConstants {
-    public static final int FRONT_GEAR_RATIO = 0;
-    public static double SETPOINT_TOLERANCE = 0.05; // Percentages
+    public static final double GEAR_RATIO = 1.0;
+    public static final double SETPOINT_TOLERANCE = 0.05; // Percentages
+    public static final double MOMENT_OF_INERTIA = 0.08;
+
+    public static final LoggedTunableNumber TOP_P = new LoggedTunableNumber("Shooter/Top P");
+    public static final LoggedTunableNumber TOP_I = new LoggedTunableNumber("Shooter/Top I");
+    public static final LoggedTunableNumber TOP_D = new LoggedTunableNumber("Shooter/Top D");
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: `kP` instead of `P`.

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java [hunk 6]
```diff
@@ -1,6 +1,62 @@
...
+                TOP_Ks.initDefault(0.0);
+                TOP_Kv.initDefault(0.0);
+                TOP_Ka.initDefault(0.0);
+
+                BOTTOM_P.initDefault(0.0);
+                BOTTOM_I.initDefault(0.0);
+                BOTTOM_D.initDefault(0.0);
+                BOTTOM_Ks.initDefault(0.0);
+                BOTTOM_Kv.initDefault(0.0);
+                BOTTOM_Ka.initDefault(0.0);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]:
```suggestion
            case REAL:
                TOP_P.initDefault(10.0);
                TOP_I.initDefault(0.0);
                TOP_D.initDefault(0.0);
                TOP_Ks.initDefault(0.0);
                TOP_Kv.initDefault(0.112);
                TOP_Ka.initDefault(0.0);

                BOTTOM_P.initDefault(10.0);
                BOTTOM_I.initDefault(0.0);
                BOTTOM_D.initDefault(0.0);
                BOTTOM_Ks.initDefault(0.0);
                BOTTOM_Kv.initDefault(0.112);
                BOTTOM_Ka.initDefault(0.0);
            case SIM:
            case REPLAY:
                TOP_P.initDefault(0.0);
                TOP_I.initDefault(0.0);
                TOP_D.initDefault(0.0);
                TOP_Ks.initDefault(0.0);
                TOP_Kv.initDefault(0.0);
                TOP_Ka.initDefault(0.0);

                BOTTOM_P.initDefault(0.0);
                BOTTOM_I.initDefault(0.0);
                BOTTOM_D.initDefault(0.0);
                BOTTOM_Ks.initDefault(0.0);
                BOTTOM_Kv.initDefault(0.0);
                BOTTOM_Ka.initDefault(0.0);
```

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    private final VelocityVoltage bottomControl = new VelocityVoltage(0);
+    private final SimpleMotorFeedforward bottomFeedForward;
+
+    public ShooterIOSim() {
+        topMotor =
+                new TalonFXSim(
+                        1, ShooterConstants.GEAR_RATIO, ShooterConstants.MOMENT_OF_INERTIA, 1);
+        bottomMotor =
+                new TalonFXSim(
+                        1, ShooterConstants.GEAR_RATIO, ShooterConstants.MOMENT_OF_INERTIA, 1);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java]: The gear of ratio and MoI might be different between the two rollers.

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import edu.wpi.first.units.Angle;
+import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.units.Velocity;
+import edu.wpi.first.wpilibj.Timer;
+import lib.motors.TalonFXSim;
+
+public class ShooterIOSim implements ShooterIO {
+    private final TalonFXSim topMotor;
+    private final TalonFXSim bottomMotor;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java:13]: Will be good to explicitly config neutral mode to "coast".

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 4 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: ad05a7a

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 5 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: 3338e24

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 3 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 8168e23

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 2 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 7f24863

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 6 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: 912225b

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java (hunk 7 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java]: 4a6531a

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 4 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: $k_V$ and all other gains should be in the same convention 🙂

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 3 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Constants should be written in `ALL_CAPS`

Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: daeb6d7

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 4 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: 0b56b1b

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java (hunk 8 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java:13]: how do i config neutral mode in a simulation?

File: src/main/java/frc/robot/Robot.java (hunk 1 above)
Comment (rakrakon) [reply src/main/java/frc/robot/Robot.java]: abf0717

File: src/main/java/frc/robot/RobotContainer.java
Comment (rakrakon) [reply src/main/java/frc/robot/RobotContainer.java:1]: 0d499e8

Comment (rakrakon) [reply src/main/java/frc/robot/RobotContainer.java:1]: like this?
709582c
2f3c79f

Comment (katzuv) [reply src/main/java/frc/robot/RobotContainer.java:1]: Wow, kudos to you 😄! Note that AFAIK unit tests are not required (@vichik123 might say otherwise), but here are a few notes:
1. What is the purpose of the `close` method?
2. Currently you set the default command and immediately after that check whether the subsystem has reached its setpoint. This process is not instantaneous. You should probably add some delay before checking.

Again, kol hakavod for the effort! I suggest you read the following article on how to write unit tests in robot code: https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/unit-testing.html

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 5 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: There are still a few places where it's `K{lowercase}` instead of `k{uppercase}`, all of them should be fixed. Note that the constant name itself doesn't have to follow this convention (there it's fine to leave all caps), but the string should be e.g. `kV`.

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java (hunk 8 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java:13]: So... now I remember you don't use Phoenix 6's `TalonFXSimState` but your own class. If you would have used Phoenix 6's class, you'd have a regular `TalonFX` object which you could configure the same way you'd configure a "real" Talon FX. Maybe this is something we can work on next off-season. For now, resolving.

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 5 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: d0fa7d9

File: src/main/java/frc/robot/RobotContainer.java
Comment (rakrakon) [reply src/main/java/frc/robot/RobotContainer.java:1]: 5f68dfa
caacb73

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 5 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: Great!

File: src/main/java/frc/robot/RobotContainer.java
Comment (katzuv) [reply src/main/java/frc/robot/RobotContainer.java:1]: I made the following changes, so the velocity setpoints passed to the simulated motors are always zero:
```diff
diff --git a/src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java b/src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java
index 8bf6314..5811346 100644                                                                                                           
--- a/src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java                                                                      
+++ b/src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java                                                                      
@@ -64,7 +64,7 @@ public class ShooterIOSim implements ShooterIO {                                                                      
                 topControl                                                                                                             
                         .withVelocity(velocity.in(Units.RotationsPerSecond))
                         .withFeedForward(
-                                topFeedForward.calculate(velocity.in(Units.RotationsPerSecond))));
+                                topFeedForward.calculate(velocity.in(Units.RotationsPerSecond) * 0)));
     }

     @Override
@@ -75,7 +75,7 @@ public class ShooterIOSim implements ShooterIO {
                         .withVelocity(velocity.in(Units.RotationsPerSecond))
                         .withFeedForward(
                                 bottomFeedForward.calculate(
-                                        velocity.in(Units.RotationsPerSecond))));
+                                        velocity.in(Units.RotationsPerSecond) * 0)));
```
The test passes, which doesn't make sense -- the motors don't move, yet the subsystem somehow reaches its setpoint.
I have an idea why this happens, but I'd like you to try and understand it.

Comment (rakrakon) [reply src/main/java/frc/robot/RobotContainer.java:1]: it looks like 
```
     class ShooterIOSim...
 43                   io.setTopVelocity(topVelocity.get());
 44                   io.setBottomVelocity(bottomVelocity.get());
```
somehow doesn't actually get called so the setpoints stay 0

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 9]
```diff
@@ -35,21 +36,30 @@ public static void initialize(ShooterIO io) {
...
+
     public Command setVelocity(Supplier<MutableMeasure<Velocity<Angle>>> velocity) {
-        return run(() -> io.setVelocity(velocity.get())).withName("Set Shooter Velocity");
+        return setVelocity(velocity, velocity);
     }
 
-    public MutableMeasure<Velocity<Angle>> getVelocity() {
-        return inputs.velocity;
+    public Command stop() {
+        return setVelocity(() -> ShooterConstants.STOP_POWER);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Stopping should always be velocity zero. But also, remember that motors have a neutral mode, which should be used instead. If you use velocity zero it will use PID to get there, and that's not what you want.

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java [hunk 10]
```diff
@@ -8,21 +8,24 @@
...
+                MutableMeasure.zero(RotationsPerSecond);
+        public MutableMeasure<Velocity<Angle>> bottomVelocity =
+                MutableMeasure.zero(RotationsPerSecond);
+        public MutableMeasure<Velocity<Angle>> topVelocitySetpoint =
+                MutableMeasure.zero(RotationsPerSecond);
+        public MutableMeasure<Velocity<Angle>> bottomVelocitySetpoint =
                 MutableMeasure.zero(RotationsPerSecond);
-        public MutableMeasure<Voltage> voltage = MutableMeasure.zero(Volts);
+        public MutableMeasure<Voltage> topVoltage = MutableMeasure.zero(Volts);
+        public MutableMeasure<Voltage> bottomVoltage = MutableMeasure.zero(Volts);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: Maybe we should just have two sets of inputs?

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: I'm not sure how that would work actually, because the auto log has no way of knowing which set of inputs is which.

File: src/test/java/frc/robot/subsystems/example/ShooterTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        ShooterIO shooterIO = new ShooterIOSim();
+        Shooter.initialize(shooterIO);
+        shooter = Shooter.getInstance();
+    }
+
+    @Test
+    public void testSetVelocity() throws InterruptedException {
+        shooter.setDefaultCommand(
+                shooter.setVelocity(() -> Units.RotationsPerSecond.of(100).mutableCopy()));
+        Thread.sleep(3000);
```
Comment (vichik123) [inline src/test/java/frc/robot/subsystems/example/ShooterTest.java]: Setting it as the default command won't do anything if the line `CommandScheduler.getInstance().run()` is not called, which is done in the robot periodic. I think just adding that before sleeping the thread would do the trick.

**Review by vichik123 — COMMENTED** [review]
Good job! (I mean the changes that @rakrakon made)

File: src/main/java/frc/robot/RobotContainer.java
Comment (vichik123) [reply src/main/java/frc/robot/RobotContainer.java:1]: > Wow, kudos to you 😄! Note that AFAIK unit tests are not required (@vichik123 might say otherwise).

Well they're always good, but since we didn't have to time to go over them in the offseason they're not a requirement. Great job though on the test!

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java (hunk 10 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: WDYM?

File: src/test/java/frc/robot/subsystems/example/ShooterTest.java (hunk 11 above)
Comment (katzuv) [reply src/test/java/frc/robot/subsystems/example/ShooterTest.java]: You can try but I don't think it'll be enough, this function runs only a single loop. I'm in contact with a WPILib developer I know (was a student of mine in the course) who wrote the documentation I linked, and will get back to you with an answer.

File: src/main/java/frc/robot/RobotContainer.java
Comment (katzuv) [reply src/main/java/frc/robot/RobotContainer.java:1]: I opened this thread but feel free to resolve it by yourselves.

**Review by katzuv — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java (hunk 10 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: Well it would be easier to have a `bottomRollerInputs` and a `topRollerInputs`, instead of doubling the number of inputs in the class.

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: Ah, I see. I think this might be a nice idea. If we can even have `RollerInputs` class that gets the roller name, our code will be more concise.
We could even create one `Roller` subsystem and instantiate it twice… but that's probably a bit overkill.

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: @rakrakon you can try if you like

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 9 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: e43c8ff
a7ae657
6abe134

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java [hunk 12]
```diff
@@ -20,6 +18,14 @@ public class ShooterIOSim implements ShooterIO {
     private final VelocityVoltage bottomControl = new VelocityVoltage(0);
     private final SimpleMotorFeedforward bottomFeedForward;
 
+    private final DutyCycleOut stop =
+            new DutyCycleOut(
+                    ShooterConstants.STOP_POWER.in(Units.RotationsPerSecond),
+                    false,
+                    true,
+                    false,
+                    false);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java]: The real Talon FX class has a convenience method called [`stopMotor()`](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/hardware/TalonFX.html#stopMotor()) that uses the [`NeutralOut`](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/controls/NeutralOut.html) control request. Is it possible to use the same method here?

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java (hunk 10 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: It looks like there is no way to pass a class instance to the autoLog
```java
package frc.robot.subsystems.shooter;

import edu.wpi.first.units.Angle;
import edu.wpi.first.units.MutableMeasure;
import edu.wpi.first.units.Velocity;
import edu.wpi.first.units.Voltage;
import org.littletonrobotics.junction.AutoLog;

import static edu.wpi.first.units.Units.RotationsPerSecond;
import static edu.wpi.first.units.Units.Volts;

@AutoLog
public class RollerInputs {
    public String rollerName;
    public MutableMeasure<Velocity<Angle>> velocity =
            MutableMeasure.zero(RotationsPerSecond);
    public MutableMeasure<Velocity<Angle>> velocitySetpoint =
            MutableMeasure.zero(RotationsPerSecond);
    public MutableMeasure<Voltage> voltage = MutableMeasure.zero(Volts);

    public RollerInputs(){

    }

    public RollerInputs(String rollerName){
        this.rollerName = rollerName;
    }

    public String getRollerName(){
        return rollerName;
    }
}

```

```java
package frc.robot.subsystems.shooter;

import static edu.wpi.first.units.Units.*;

import edu.wpi.first.units.*;
import org.littletonrobotics.junction.AutoLog;

public interface ShooterIO {
    ShooterInputsAutoLogged inputs = new ShooterInputsAutoLogged();

    void setTopVelocity(MutableMeasure<Velocity<Angle>> velocity);

    void setBottomVelocity(MutableMeasure<Velocity<Angle>> velocity);

    void stop();

    /** Update the inputs of the Shooter */
    void updateInputs();

    @AutoLog
    class ShooterInputs {
        RollerInputs topRoller = new RollerInputs("TopRoller");
        RollerInputs bottomRoller = new RollerInputs("BottomRoller");
    }
}
```

I get an error because it can't do table.put to the type

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: 1. Note that you should include the language name in a code block, that way Markdown will syntax highlight your code. I edited your comment, you can see what I did in the comment history.
2. I'm gonna need to see the full error, but you might need to duplicate some code and write two inputs classes, one for each roller. You can create an issue about that in the AdvantageKit repository, I guess that's just not a feature they thought of.

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java (hunk 12 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java]: I don't think so

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java (hunk 10 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]:
```
D:\Users\rakra\robot-2024\build\generated\sources\annotationProcessor\java\main\frc\robot\subsystems\shooter\ShooterInputsAutoLogged.java:11: error: no suitable method found for put(String,RollerInputs)
    table.put("TopRoller", topRoller);
         ^
    method LogTable.put(String,LogValue) is not applicable
      (argument mismatch; RollerInputs cannot be converted to LogValue)
    method LogTable.put(String,byte[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to byte[])
    method LogTable.put(String,boolean) is not applicable
      (argument mismatch; RollerInputs cannot be converted to boolean)
    method LogTable.put(String,int) is not applicable
      (argument mismatch; RollerInputs cannot be converted to int)
    method LogTable.put(String,long) is not applicable
      (argument mismatch; RollerInputs cannot be converted to long)
    method LogTable.put(String,float) is not applicable
      (argument mismatch; RollerInputs cannot be converted to float)
    method LogTable.put(String,double) is not applicable
      (argument mismatch; RollerInputs cannot be converted to double)
    method LogTable.put(String,String) is not applicable
      (argument mismatch; RollerInputs cannot be converted to String)
    method LogTable.<E>put(String,E) is not applicable
      (inference variable E has incompatible bounds
        upper bounds: Enum<E>
        lower bounds: RollerInputs)
    method LogTable.<U>put(String,Measure<U>) is not applicable
      (cannot infer type-variable(s) U
        (argument mismatch; RollerInputs cannot be converted to Measure<U>))
    method LogTable.put(String,boolean[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to boolean[])
    method LogTable.put(String,int[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to int[])
    method LogTable.put(String,long[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to long[])
    method LogTable.put(String,float[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to float[])
    method LogTable.put(String,double[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to double[])
    method LogTable.put(String,String[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to String[])
    method LogTable.<T#1>put(String,Struct<T#1>,T#1) is not applicable
      (cannot infer type-variable(s) T#1
        (actual and formal argument lists differ in length))
    method LogTable.<T#2>put(String,Struct<T#2>,T#2...) is not applicable
      (cannot infer type-variable(s) T#2
        (argument mismatch; RollerInputs cannot be converted to Struct<T#2>))
    method LogTable.<T#3,MessageType>put(String,Protobuf<T#3,MessageType>,T#3) is not applicable
      (cannot infer type-variable(s) T#3,MessageType
        (actual and formal argument lists differ in length))
    method LogTable.<T#4>put(String,T#4) is not applicable
      (inference variable T#4 has incompatible bounds
        upper bounds: WPISerializable
        lower bounds: RollerInputs)
    method LogTable.<T#5>put(String,T#5...) is not applicable
      (inference variable T#5 has incompatible bounds
        upper bounds: StructSerializable
        lower bounds: RollerInputs)
  where E,U,T#1,T#2,T#3,MessageType,T#4,T#5 are type-variables:
    E extends Enum<E> declared in method <E>put(String,E)
    U extends Unit<U> declared in method <U>put(String,Measure<U>)
    T#1 extends Object declared in method <T#1>put(String,Struct<T#1>,T#1)
    T#2 extends Object declared in method <T#2>put(String,Struct<T#2>,T#2...)
    T#3 extends Object declared in method <T#3,MessageType>put(String,Protobuf<T#3,MessageType>,T#3)
    MessageType extends ProtoMessage<?> declared in method <T#3,MessageType>put(String,Protobuf<T#3,MessageType>,T#3)
    T#4 extends WPISerializable declared in method <T#4>put(String,T#4)
    T#5 extends StructSerializable declared in method <T#5>put(String,T#5...)
```

File: src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java (hunk 12 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIOSim.java]: Okay, so just remember to use it in the real implementation.

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java (hunk 10 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: Yep:
```
no suitable method found for put(String,RollerInputs)
```
Not supported for classes. But making two inputs classes should work.

Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: fe871ab
0393153
d7058f6
e040319

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: That's a better solution than I thought of! Did you check and see if you really see two different inputs sets in AdvantageScope?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 9 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: e040319
